### ISSUE TITLE
[C++4OpenCL] Improve diagnostics of kernel param types

### ISSUE THREAD

**Issue by AnastasiaStulova** [llvm/llvm-project#58590]
C++ for OpenCL made some diagnostics for kernel param types more conservative and sometimes even incorrect. For example OpenCL C didn't disallow use of forward declared types but in C++ for OpenCL they are not allowed due to standard layout and other C++ specific requirement failing.

One of the biggest problems though that templated types are not diagnosed correctly when passed by reference or pointer due to lazy template instantiation.

One idea explained in https://reviews.llvm.org/D134445 is to turn such incorrect diagnostics into warning instead with possibility to disable the warning by a flag. The warning could inform that diagnosing type is not possible and that developers need to do manual checking of their correctness. This is better than preventing using the types due to inability of correct diagnosing.

**Comment by AnastasiaStulova** [comment]
Note the following comment suggesting different approach: https://reviews.llvm.org/D134445#3920448

**Comment by bmanga** [comment]
I think I stumbled on this issue as well (through clspv): https://github.com/google/clspv/issues/1053
If I understand correctly for now the best solution is to still add 
`#pragma OPENCL EXTENSION __cl_clang_non_portable_kernel_param_types : enable` before the kernel declarations, correct?

**Comment by AnastasiaStulova** [comment]
Correct, this is the best solution we have at the moment before we improve the false detection of unsafe types or at least turn those into warnings.

**Comment by bmanga** [comment]
Thank you :)
